# Web APIs

## 1. Consuming Web APIs

(Partially adapted from https://www.dataquest.io/blog/python-api-tutorial/)

* https://randomuser.me/ is a random user generator 
*	It has an API. Go to https://api.randomuser.me/ with your browser
*	You get a JSON (JavaScript Object Notation) back. A JSON is similar to a Python dictionary.  
*  Refresh the browser
*	You can add query parameters. They are added to the URL with a ? . You then add then the parameter name = the value. More the one parameter are connected with a & e.g.
	http://api.open-notify.org/iss-pass.json?lat=37.78&lon=-122.41

*	 Go the documentation https://randomuser.me/documentation and check how you can get multiple users and to specify constraints on the output
*	Go to your browser and add parameters to https://api.randomuser.me/ so that you get 5 results of only males from the US

**Enter the URL of your solution:**

In [17]:
# TODO
url = None

* You can also get the data from the command line. Open the command line and write

```bash
curl -s https://api.randomuser.me/
```

You can also run Bash commands directly in your Jupyter Notebook with !:

In [2]:
!curl -s https://api.randomuser.me/

{"results":[{"gender":"male","name":{"title":"mr","first":"zenon","last":"freyer"},"location":{"street":"kirchgasse 1","city":"vetschau/spreewald","state":"rheinland-pfalz","postcode":89227,"coordinates":{"latitude":"-45.3792","longitude":"-141.9297"},"timezone":{"offset":"+3:00","description":"Baghdad, Riyadh, Moscow, St. Petersburg"}},"email":"zenon.freyer@example.com","login":{"uuid":"f3888eae-4439-4fe6-950b-142c611be6e3","username":"angrypanda167","password":"fields","salt":"spnM74zL","md5":"2f3b83652776e3887b1308798905609c","sha1":"cd97c4ca4fd34f36d8b6a894785fa9905e074522","sha256":"69e2e7e117a151d3b1fb3ac3d18ff15a40076c900a1201d7f03ad2806143baa3"},"dob":{"date":"1960-06-09T13:15:46Z","age":58},"registered":{"date":"2004-08-15T18:11:38Z","age":14},"phone":"0401-6949282","cell":"0176-0056449","id":{"name":"","value":null},"picture":{"large":"https://randomuser.me/api/portraits/men/80.jpg","medium":"https://randomuser.me/api/portraits/med/men/80.jpg","thumbnail":"https://randomuser.

* Import the two libraries requests and json. You can find the documentation for the requests package here:
 http://www.python-requests.org/en/latest/ 

In [4]:
import requests
import json

* With the requests package you can call a Web API with the URL and the method get

In [18]:
response = requests.get("https://api.randomuser.me/")

<Response [200]>

* Print the status code of the request

In [6]:
print(response.status_code)

200


**The meaning of the status codes are:**
* 200: everything went okay, and the result has been returned (if any)
* 301: the server is redirecting you to a different endpoint. This can happen when a company switches domain names, or an endpoint name is changed.
* 401: the server thinks you're not authenticated. This happens when you don't send the right credentials to access an API.
* 400: the server thinks you made a bad request. This can happen when you don't send along the right data, among other things.
* 403: the resource you're trying to access is forbidden – you don't have the right permissions to see it.
* 404: the resource you tried to access wasn't found on the server.


* Get with the request method 10 results of only males from the US. You can specify the query parameters with a Python dictionary like this:
```python
parameters = {"lat": 37.78, "lon": -122.41}
```

In [7]:
# TODO:

And pass the parameter to the request like this
```python
response = requests.get("http://api.open-notify.org/iss-pass.json", params=parameters)
```

Alternatively you could build the URL also by yourself

In [8]:
# TODO:

* You can show the result of the request with the method text

In [9]:
response.text

'{"results":[{"gender":"female","name":{"title":"miss","first":"eva","last":"russell"},"location":{"street":"7021 woodland st","city":"sunshine coast","state":"victoria","postcode":7121,"coordinates":{"latitude":"-80.0569","longitude":"66.5126"},"timezone":{"offset":"+3:30","description":"Tehran"}},"email":"eva.russell@example.com","login":{"uuid":"5c492fcd-953d-4aff-8c0b-68d5733c4948","username":"sadgorilla605","password":"1226","salt":"vn2P09cY","md5":"bf6c6dc878b4f205aaac324aebb8b13c","sha1":"ff64770502a28c308879bfaddd78c5aa625743e3","sha256":"d52cc2200922dff80deb5cd1b1f8f570a039707fed5574304a3cd7c2cb075523"},"dob":{"date":"1990-11-04T02:46:32Z","age":28},"registered":{"date":"2008-03-23T02:52:34Z","age":11},"phone":"02-2735-3525","cell":"0419-415-480","id":{"name":"TFN","value":"411188414"},"picture":{"large":"https://randomuser.me/api/portraits/women/47.jpg","medium":"https://randomuser.me/api/portraits/med/women/47.jpg","thumbnail":"https://randomuser.me/api/portraits/thumb/women

* You can convert the data from JSON to a Python dictionary with the package JSON

In [10]:
data = json.loads(response.text)

* Check the type of variable data

In [30]:
# TODO

In [28]:
print(data)

{'results': [{'gender': 'female', 'name': {'title': 'miss', 'first': 'eva', 'last': 'russell'}, 'location': {'street': '7021 woodland st', 'city': 'sunshine coast', 'state': 'victoria', 'postcode': 7121, 'coordinates': {'latitude': '-80.0569', 'longitude': '66.5126'}, 'timezone': {'offset': '+3:30', 'description': 'Tehran'}}, 'email': 'eva.russell@example.com', 'login': {'uuid': '5c492fcd-953d-4aff-8c0b-68d5733c4948', 'username': 'sadgorilla605', 'password': '1226', 'salt': 'vn2P09cY', 'md5': 'bf6c6dc878b4f205aaac324aebb8b13c', 'sha1': 'ff64770502a28c308879bfaddd78c5aa625743e3', 'sha256': 'd52cc2200922dff80deb5cd1b1f8f570a039707fed5574304a3cd7c2cb075523'}, 'dob': {'date': '1990-11-04T02:46:32Z', 'age': 28}, 'registered': {'date': '2008-03-23T02:52:34Z', 'age': 11}, 'phone': '02-2735-3525', 'cell': '0419-415-480', 'id': {'name': 'TFN', 'value': '411188414'}, 'picture': {'large': 'https://randomuser.me/api/portraits/women/47.jpg', 'medium': 'https://randomuser.me/api/portraits/med/women/

* *pretty-print* (pprint) prints complex data structures like dictionary prettier. 

In [26]:
from pprint import pprint
pprint(data)

{'info': {'page': 1,
          'results': 1,
          'seed': 'c55efca4e75f0a0f',
          'version': '1.2'},
 'results': [{'cell': '0419-415-480',
              'dob': {'age': 28, 'date': '1990-11-04T02:46:32Z'},
              'email': 'eva.russell@example.com',
              'gender': 'female',
              'id': {'name': 'TFN', 'value': '411188414'},
              'location': {'city': 'sunshine coast',
                           'coordinates': {'latitude': '-80.0569',
                                           'longitude': '66.5126'},
                           'postcode': 7121,
                           'state': 'victoria',
                           'street': '7021 woodland st',
                           'timezone': {'description': 'Tehran',
                                        'offset': '+3:30'}},
              'login': {'md5': 'bf6c6dc878b4f205aaac324aebb8b13c',
                        'password': '1226',
                        'salt': 'vn2P09cY',
                      

* Pretty Print only the results of the data

* Loop through the dictionary and print all first names

In [12]:
# TODO

* Get all astronauts who are right now in space. You get the information about the Web APU from here  http://open-notify.org/Open-Notify-API/People-In-Space/ 

In [13]:
# TODO

* Print the number of people that are right now in space

In [15]:
# TODO

* Print the names of all astronauts 

In [33]:
# TODO

* A lot of Web APIs require a key for interacting with them (like Twitter, Facebook, …). You find at http://www.python-requests.org/en/latest/user/authentication/ more information for Authentication for Web APIs with the request package
* There are also special Python packages for interacting with services. E.g. for Twitter: http://www.tweepy.org/ or  https://github.com/bear/python-twitter 

See e.g. http://socialmedia-class.org/twittertutorial.html for a tutorial

## 2. Creating a Web API

* Create a folder `webapi` and change into it.
* Create in the `webapi` folder a `Dockerfile` with the following content:

```bash
# Use an official Python runtime as a parent image
FROM python:3.7-slim

# Set the working directory to /app
WORKDIR /app

# Copy the current directory contents into the container at /app
COPY app/ /app

# Install any needed packages specified in requirements.txt
RUN pip install --trusted-host pypi.python.org -r requirements.txt

# Make port 80 available to the world outside this container
EXPOSE 80

# Run app.py when the container launches
CMD ["python", "app.py"]
```

* We can also use Docker compose with just one service. Create in your `webapi` folder a `docker-compose.yml` file:

```yaml
version: '3'
services:
  api:
    build: .
    ports:
      - "5000:80"
    restart: always
    volumes:
      - ./app:/app
```

* We will build a web API with `Flask` (http://flask.pocoo.org/) . Create a `requirements.txt` file in the `app` folder. Here we can specify all python `pip` packages that we need:
```bash
Flask
```

* Create the `app.py` file in the `app` folder:

```python
from flask import Flask
from flask import request, jsonify

app = Flask(__name__)

courses = [
    {'id': 0,
     'title': 'Data Science',
     'professor': 'Markus Löcher',
     'semester': '1'},
    {'id': 1,
     'title': 'Data Warehousing',
     'professor': 'Roland M. Mueller',
     'semester': '1'},
    {'id': 2,
     'title': 'Business Process Management',
     'professor': 'Frank Habermann',
     'semester': '1'},
    {'id': 3,
     'title': 'Stratigic Issues of IT',
     'professor': 'Sven Pohland',
     'semester': '1'},
    {'id': 4,
     'title': 'Text, Web and Social Media Analytics Lab',
     'professor': 'Markus Löcher',
     'semester': '2'},
    {'id': 5,
     'title': 'Enterprise Architectures for Big Data',
     'professor': 'Roland M. Mueller',
     'semester': '2'},
    {'id': 6,
     'title': 'Business Process Integration Lab',
     'professor': 'Frank Habermann',
     'semester': '2'},
    {'id': 7,
     'title': 'IT-Security and Privacy',
     'professor': 'Dennis Uckel',
     'semester': '2'},
    {'id': 8,
     'title': 'Research Methods',
     'professor': 'Marcus Birkenkrahe',
     'semester': '2'},
]

@app.route('/api/v1/courses/all', methods=['GET'])
def api_all():
    return jsonify(courses)

@app.route('/api/v1/courses', methods=['GET'])
def api_id():
    # Check if an ID was provided as part of the URL.
    # If ID is provided, assign it to a variable.
    # If no ID is provided, display an error in the browser.
    if 'id' in request.args:
        id = int(request.args['id'])
    else:
        return "Error: No id field provided. Please specify an id."

    # Create an empty list for our results
    results = []

    # Loop through the data and match results that fit the requested ID.
    # IDs are unique, but other fields might return many results
    for course in courses:
        if course['id'] == id:
            results.append(course)

    # Use the jsonify function from Flask to convert our list of
    # Python dictionaries to the JSON format.
    return jsonify(results)

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=80, debug=True)
```

* Open http://localhost:5000/api/v1/courses/all in a browser
* Open http://localhost:5000/api/v1/courses?id=5 in a browser


* Use your API in Python here in the Jupyter Notebook and print all names of all courses 

In [1]:
# TODO

* Add the possibility to find courses based on the semester
* Use your API in Python and print all names of all courses in the second semester

In [2]:
# TODO

* Add a function that can convert Fahrenheit to Celsius 
* Call you API and get the Celsius value for 100°F

In [4]:
# TODO